src/wx/widgets/sdljoy.cpp

```diff
@@ -38,420 +39,489 @@ static int16_t axisval(int16_t x)
+                    e.type, e.jaxis.axis, AxisValueToDirection(e.jaxis.value));
             }
-            else {
-                for (uint8_t but = 0; but < SDL_JoystickNumButtons(joy.second.dev); but++) {
-                    auto last_state = joy.second.button[but];
-                    auto state      = SDL_JoystickGetButton(joy.second.dev, but);
-
-                    if (last_state != state) {
-                        CreateAndSendEvent(joy.first, WXSDLJOY_BUTTON, but, state, last_state);
 
-                        joy.second.button[but] = state;
-
-                        wxLogDebug("POLLED SDL_JOYBUTTON: joy:%d but:%d val:%d prev_val:%d", joy.first, but, state, last_state);
-                    }
-                }
-
-                for (uint8_t axis = 0; axis < SDL_JoystickNumAxes(joy.second.dev); axis++) {
-                    auto val      = axisval(SDL_JoystickGetAxis(joy.second.dev, axis));
-                    auto prev_val = joy.second.axis[axis];
-
-                    if (val != prev_val) {
-                        CreateAndSendEvent(joy.first, WXSDLJOY_AXIS, axis, val, prev_val);
-
-                        joy.second.axis[axis] = val;
-
-                        wxLogDebug("POLLED SDL_JOYAXISMOTION: joy:%d axis:%d val:%d prev_val:%d", joy.first, axis, val, prev_val);
-                    }
-                }
-            }
+            got_event = true;
+            break;
         }
-    }
-}
 
-void wxSDLJoy::ConnectController(uint8_t joy)
-{
-    SDL_Joystick* js_dev = nullptr;
-    bool is_gc;
-
-    if ((is_gc = SDL_IsGameController(joy))) {
-        auto dev = SDL_GameControllerOpen(joy);
+        case SDL_JOYHATMOTION:
+        {
+            SDL_JoystickID joy_id = e.jhat.which;
+            auto joy_state = joystick_states_.find(joy_id);
+            if (joy_state != joystick_states_.end()) {
+                // Do not process joystick events for game controllers.
+                if (joy_state->second->IsGameController())
+                    break;
 
-        if (dev) {
-            joystate[joy].dev = dev;
+                joy_state->second->ProcessEvent(
+                    e.type, e.jhat.hat, AxisValueToDirection(e.jhat.value));
+            }
 
-            js_dev = SDL_GameControllerGetJoystick(dev);
+            got_event = true;
+            break;
         }
-    }
-    else {
-        if ((js_dev = SDL_JoystickOpen(joy)))
-            joystate[joy].dev = js_dev;
-    }
 
-    if (js_dev) {
-        auto instance = SDL_JoystickInstanceID(js_dev);
+        case SDL_JOYDEVICEADDED:
+        {
+            int sdl_index = e.jdevice.which;
+            // Always remap all controllers.
+            RemapControllers();
+            got_event = true;
+            break;
+        }
 
-        instance_map[instance] = &(joystate[joy]);
+        case SDL_JOYDEVICEREMOVED:
+        {
+            SDL_JoystickID joy_id = e.jdevice.which;
+            if (add_all_joysticks_)
```

Comment:
The variable is used to differentiate between adding a select list of joysticks or listening to all joysticks. I renamed it to `poll_all_joysticks_` since it was confusing.